### PR TITLE
Pass explicit state, not a success boolean

### PR DESCRIPTION
**Pull request by cansjt**
This change will allow mark tasks to states other than success (e.g. skipped).

The new API should be backward compatible with the existing one and provides explicit deprecation warnings.

After the deprecation period expired, the changes to clean the code should be straight forward: remove some formal parameters, checks and deprecation warnings.

related: #17482 

TODO:
- [ ] we should probably be more defensive in terms of the states accepted by the different functions & methods.
- [ ] depending on how public or final not some of the functions modified are, we might not need to keep certain deprecation warnings.
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -571,9 +595,24 @@ def log_url(self):
...
+        base_url = conf.get('webserver', 'BASE_URL')
+        return base_url + (
+            "/confirm"
+            f"?task_id={self.task_id}"
+            f"&dag_id={self.dag_id}"
+            f"&execution_date={iso}"
+            "&upstream=false"
+            "&downstream=false"
+            f"&state={State.SKIPPED}"
+        )
```
Comment (cansjt) [inline airflow/models/taskinstance.py:611]: Would this be useful ?

File: airflow/cli/cli_parser.py [hunk 2]
```diff
@@ -219,7 +219,16 @@ def _check(value):
 
 # backfill
 ARG_MARK_SUCCESS = Arg(
-    ("-m", "--mark-success"), help="Mark jobs as succeeded without running them", action="store_true"
+    ("-m", "--mark-success"),
+    action="store_const",
+    const='success',
+    dest='mark_as',
```
Comment (cansjt) [inline airflow/cli/cli_parser.py:227]: If the definition of those arguments are considered public API, e.g. available for reuse in plugins, then this change is not backward compatible.

But as it is, it avoids having to add boilerplate code to check the arguments values post parsing. Minimizing the changes needed in `airflow/cli/commands/dag_command.py` and `airflow/cli/commands/task_command.py`.

File: airflow/cli/cli_parser.py [hunk 3]
```diff
@@ -219,7 +221,17 @@ def _check(value):
 
 # backfill
 ARG_MARK_SUCCESS = Arg(
-    ("-m", "--mark-success"), help="Mark jobs as succeeded without running them", action="store_true"
+    ("-m", "--mark-success"),
+    action="store_const",
+    const=TaskInstanceState.SUCCESS,
+    dest='mark_as',
+    help="Mark jobs as succeeded without running them " "(deprecated, use --mark-as=success instead)",
```
Comment (uranusjr) [inline airflow/cli/cli_parser.py]: Need to fix this (Black 🤦)

File: airflow/executors/base_executor.py [hunk 4]
```diff
@@ -97,6 +100,15 @@ def queue_task_instance(
         cfg_path: Optional[str] = None,
     ) -> None:
         """Queues task instance."""
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
+            )
+            if mark_as is not None:
+                raise AirflowException('You cannot use both `mark_as` and `mark_success` arguments')
```
Comment (uranusjr) [inline airflow/executors/base_executor.py]:
```suggestion
                raise TypeError('You cannot use both `mark_as` and `mark_success` arguments')
```

Incorrect function signature should be type error.

File: airflow/cli/cli_parser.py (hunk 2 above)
Comment (ashb) [reply airflow/cli/cli_parser.py:227]: No, these consts/objects do not really form part of the public API (the CLI args do, but I'm not worried about breaking compat with this sort of change)

File: airflow/jobs/backfill_job.py [hunk 5]
```diff
@@ -146,8 +148,15 @@ def __init__(
         :type start_date: datetime.datetime
         :param end_date: end date for the backfill date range.
         :type end_date: datetime.datetime
+        :param mark_as: Mark the task as being in the given state.
+            Value must be one of :class:`.TaskInstanceState`.
+        :type mark_as: .TaskInstanceState
         :param mark_success: flag whether to mark the task auto success.
-        :type mark_success: bool
+        :type mark_success: airflow.utils.state.State
+           .. deprecated: 2.1.3
```
Comment (ashb) [inline airflow/jobs/backfill_job.py:156]:
```suggestion
           .. deprecated: 2.2.0
```

File: airflow/models/dag.py [hunk 6]
```diff
@@ -1939,7 +1940,13 @@ def run(
         :type start_date: datetime.datetime
         :param end_date: the end date of the range to run
         :type end_date: datetime.datetime
+        :param mark_as: Mark the task as being in the given state.
+        :type mark_as: .TaskInstanceState
         :param mark_success: True to mark jobs as succeeded without running them
+           .. deprecated: 2.1.3
```
Comment (ashb) [inline airflow/models/dag.py:1946]:
```suggestion
           .. deprecated: 2.2.0
```

File: airflow/models/taskinstance.py [hunk 7]
```diff
@@ -505,7 +515,13 @@ def generate_command(
         :type task_id: str
         :param execution_date: Execution date for the task
         :type execution_date: datetime
+        :param mark_as: Mark the task as being in the given state.
+        :type mark_as: .TaskInstanceState
         :param mark_success: Whether to mark the task as successful
+           .. deprecated: 2.1.3
```
Comment (ashb) [inline airflow/models/taskinstance.py:521]:
```suggestion
           .. deprecated: 2.2.0
```

File: airflow/models/taskinstance.py [hunk 8]
```diff
@@ -1039,7 +1079,13 @@ def check_and_change_state_before_execution(
         :type ignore_task_deps: bool
         :param ignore_ti_state: Disregards previous task instance state
         :type ignore_ti_state: bool
+        :param mark_as: Don't run the task, mark it as being in the given state.
+        :type mark_as: .TaskInstanceState
         :param mark_success: Don't run the task, mark its state as success
+           .. deprecated: 2.1.3
```
Comment (ashb) [inline airflow/models/taskinstance.py:1085]:
```suggestion
           .. deprecated: 2.2.0
```

File: airflow/models/taskinstance.py [hunk 9]
```diff
@@ -1052,6 +1098,14 @@ def check_and_change_state_before_execution(
         :return: whether the state was changed to running or not
         :rtype: bool
         """
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/taskinstance.py:1104]:
```suggestion
                DeprecationWarning,
                stacklevel=3
```
(+1 for the `@provide_session`)

File: airflow/models/taskinstance.py [hunk 10]
```diff
@@ -1178,7 +1233,13 @@ def _run_raw_task(
         completion and runs any post-execute callbacks. Meant to be called
         only after another function changes the state to running.
 
+        :param mark_as: Don't run the tasks, mark it with the given state.
+        :type: mark_as: .TaskInstanceState
         :param mark_success: Don't run the task, mark its state as success
+           .. deprecated: 2.1.3
```
Comment (ashb) [inline airflow/models/taskinstance.py:1239]:
```suggestion
           .. deprecated: 2.2.0
```

File: airflow/models/taskinstance.py [hunk 11]
```diff
@@ -1187,6 +1248,14 @@ def _run_raw_task(
         :param session: SQLAlchemy ORM Session
         :type session: Session
         """
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/taskinstance.py:1254]:
```suggestion
                DeprecationWarning,
                stacklevel=3,
```

File: airflow/models/taskinstance.py [hunk 12]
```diff
@@ -1390,20 +1459,29 @@ def run(
...
         test_mode: bool = False,
         job_id: Optional[str] = None,
         pool: Optional[str] = None,
         session=None,
     ) -> None:
         """Run TaskInstance"""
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/taskinstance.py:1473]:
```suggestion
                DeprecationWarning,
                stacklevel=3,
```

**Review by ashb — COMMENTED** [review]
Nice one, and good job on the compat layer too!

File: airflow/executors/base_executor.py [hunk 13]
```diff
@@ -97,6 +99,15 @@ def queue_task_instance(
         cfg_path: Optional[str] = None,
     ) -> None:
         """Queues task instance."""
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/executors/base_executor.py:105]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/executors/celery_kubernetes_executor.py [hunk 14]
```diff
@@ -105,20 +109,28 @@ def queue_task_instance(
         cfg_path: Optional[str] = None,
     ) -> None:
         """Queues task instance via celery or kubernetes executor"""
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/executors/celery_kubernetes_executor.py:115]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/executors/debug_executor.py [hunk 15]
```diff
@@ -100,6 +103,14 @@ def queue_task_instance(
         cfg_path: Optional[str] = None,
     ) -> None:
         """Queues task instance with empty command because we do not need it."""
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/executors/debug_executor.py:109]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/jobs/backfill_job.py [hunk 16]
```diff
@@ -169,11 +178,19 @@ def __init__(
         :param args:
         :param kwargs:
         """
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/jobs/backfill_job.py:184]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/jobs/local_task_job.py [hunk 17]
```diff
@@ -49,12 +50,21 @@ def __init__(
...
+        mark_success: Optional[bool] = None,
         pickle_id: Optional[str] = None,
         pool: Optional[str] = None,
         *args,
         **kwargs,
     ):
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/jobs/local_task_job.py:63]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/models/baseoperator.py [hunk 18]
```diff
@@ -1252,15 +1253,25 @@ def run(
...
         ignore_ti_state: bool = False,
-        mark_success: bool = False,
+        mark_as: Optional[TaskInstanceState] = None,
+        mark_success: Optional[bool] = None,
     ) -> None:
         """Run a set of task instances for a date range."""
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/baseoperator.py:1263]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/models/dag.py [hunk 19]
```diff
@@ -1967,6 +1974,14 @@ def run(
         :type: bool
 
         """
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/dag.py:1980]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/models/taskinstance.py [hunk 20]
```diff
@@ -440,6 +441,14 @@ def command_as_list(
         installed. This command is part of the message sent to executors by
         the orchestrator.
         """
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/taskinstance.py:447]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

File: airflow/models/taskinstance.py [hunk 21]
```diff
@@ -536,10 +552,18 @@ def generate_command(
         :return: shell command that can be used to run the task instance
         :rtype: list[str]
         """
+        if mark_success is not None:
+            warnings.warn(
+                'The argument `mark_success` has been deprecated. Use `state=State.SUCCESS` instead.',
+                DeprecationWarning,
```
Comment (ashb) [inline airflow/models/taskinstance.py:558]:
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

**Comment by ashb** [comment]
Looks great to me once the `stacklevel`s are  added in.

**Pull request closed without merging**